nixos/modules/services/web-apps/nextcloud.xml

```diff
@@ -0,0 +1,91 @@
+<chapter xmlns="http://docbook.org/ns/docbook"
+         xmlns:xlink="http://www.w3.org/1999/xlink"
+         xmlns:xi="http://www.w3.org/2001/XInclude"
+         version="5.0"
+         xml:id="module-services-nextcloud">
+ <title>Nextcloud</title>
+
+ <para>
+  <link xlink:href="https://nextcloud.com/">Nextcloud</link> is an open-source, self-hostable cloud
+  platform. The server setup can be automated using
+  <link linkend="opt-services.nextcloud.enable">services.nextcloud</link>. A desktop client is packaged
+  at <literal>pkgs.nextcloud-client</literal>.
+ </para>
+
+ <section xml:id="module-services-nextcloud-basic-usage">
+  <title>Basic usage</title>
+  <para>
+   Nextcloud is a PHP-based application which requires an HTTP server
+   (<literal><link linkend="opt-services.nextcloud.enable">services.nextcloud</link></literal> optionally supports
+   <literal><link linkend="opt-services.nginx.enable">services.nginx</link></literal>) and a database
+   (it's recommended to use <literal><link linkend="opt-services.postgresql.enable">services.postgresql</link></literal>).
+  </para>
+  <para>
+    A very basic configuration may look like this:
+<programlisting>{ pkgs, ... }:
+{
+  services.nextcloud = {
+    <link linkend="opt-services.nextcloud.enable">enable</link> = true;
+    <link linkend="opt-services.nextcloud.hostName">hostName</link> = "nextcloud.tld";
+    <link linkend="opt-services.nextcloud.nginx.enable">nginx.enable</link> = true;
+    config = {
+      <link linkend="opt-services.nextcloud.config.dbtype">dbtype</link> = "pgsql";
+      <link linkend="opt-services.nextcloud.config.dbpass">dbpass</link> = "certainly-very-secret";
```

Comment:
Instead of advertising a user/password authentication, I think we should advertise the unix-authentication version of the configuration in the example, where the nextcloud user is authenticated by unix socket, as it in particular avoids having a “secret” in plaintext in the store.

Reply:
BTW, forgot to write down the unix-authentication version of the configuration: mine looks like this (including only the bits that diverge from the code you wrote):
```nix
{
  # ...
  services.nextcloud.config = {
    dbtype = "pgsql";
    dbname = "nextcloud";
    dbuser = "nextcloud";
    dbhost = "/tmp"; # nextcloud will add /.s.PGSQL.5432 by itself
    adminuser = "root";
    adminpassFile = "/root/nextcloud.adminpass";
  };
  services.postgresql.initialScript = pkgs.writeText "psql-init" ''
    CREATE ROLE nextcloud WITH LOGIN;
    CREATE DATABASE nextcloud WITH OWNER nextcloud;
  '';
}
```

Reply:
Oh indeed! I had forgotten this part:
```nix
{
  # ...
  systemd.services."nextcloud-setup" = {
    requires = ["postgresql.service"];
    after = ["postgresql.service"];
  };
  # ...
}
```
I wonder whether it should be all turned on by a helper option like `services.nextcloud.postgresql = "local";` or similar.

Reply:
Actually nevermind my idea of turning this all on: `initialScript` can't be automatically turned on as IIRC it's not assumed to be idempotent, and thus the user may have to run manual steps for setup if they already had a postgresql database.

Reply:
btw I updated the documentation, in fact you're right, we shouldn't promote user/password auth :)

I kept the `initialScript` example, it's IMHO the easiest way for a setup, in case of an existing database you'd have to do this manually right? In the end Nix doesn't work for user data, so there's actually no way around,right?

Reply:
Thanks! And yes, unfortunately this appears to be a limitation of the `postgresql` module, that doesn't allow for declaratively defining databases and users. For an example of nix handling user content there's `services.openldap.declarativeContents`, but here we'd want something lighter, like `services.mysql.ensureDatabases` / `services.mysql.ensureUsers`.

Anyway, this is quite unrelated to this PR now :)

Reply:
I think @globin was working on a solution for declaratively defining that, similar to how it's allowed in the mysql module.

Reply:
You mean something similar to `ensureDatabases`/`ensureUsers`?
That's definetely a good idea, but I think that it's sufficiently documented here (the option description warns about `initialScript`) and I added references to the used NixOS options where possible.

Reply:
Agree with both your points :)